File: 2024/Day04/day4.py
#!/usr/bin/env python3
import sys

with open(sys.argv[1], 'r') as file:
    rows = [r.strip() for r in file.readlines()]


directions = {
    'u' : (-1, 0),
    'ur': (-1, 1),
    'r' : (0, 1),
    'dr': (1, 1),
    'd' : (1, 0),
    'dl': (1, -1),
    'l' : (0, -1),
    'ul': (-1, -1)
}

h, w = len(rows), len(rows[0])
chars1 = ['M','A','S']
directions2 = {k: directions[k] for k in ['ur', 'dr', 'dl', 'ul']}
matches2 = [['M', 'M', 'S', 'S'],
            ['S', 'M', 'M', 'S'],
            ['S', 'S', 'M', 'M'],
            ['M', 'S', 'S', 'M']]
cnt1 = 0
cnt2 = 0

for i, row in enumerate(rows):
    for j, ch in enumerate(row):
        if ch == 'X':

            for (direction, (di, dj)) in directions.items():
                flag = True
                ii, jj = i + di, j + dj
                for a, char in enumerate(chars1):
                    if  (0 <= ii < h) and (0 <= jj < w) and rows[ii][jj] == char:
                        ii += di
                        jj += dj
                        continue
                    else:
                        flag = False
                        break
                if flag:
                    cnt1 += 1

        if ch == 'A':

            tmp = []
            for (direction, (di, dj)) in directions2.items():
                ii, jj = i + di, j + dj
                if  (0 <= ii < h) and (0 <= jj < w):
                    tmp.append(rows[ii][jj])
            
            if tmp in matches2:
                cnt2 += 1

print(f'Part 1: {cnt1}')
print(f'Part 2: {cnt2}')

File: 2024/Day04/day4.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <sstream>
#include <string>

using namespace std;

int main() {
    ifstream f("input.txt");
    if (!f.is_open()) {
        cerr << "Error: Unable to open file." << endl;
        return 1;
    }
    vector< vector < char > > rows;
    vector< char > row;
    string line;
    while (getline(f, line)) {
        row.assign(line.begin(),line.end());
        rows.push_back(row);
    }

    const vector<vector<int> > directions = {{-1, 0}, {-1, 1}, {0, 1}, {1, 1}, {1, 0}, {1, -1}, {0, -1}, {-1, -1}};
    const vector<vector<int> > directions2 = {{-1, 1}, {1, 1}, {1, -1}, {-1, -1}};
    int h = rows.size(), w = (h > 0) ? rows[0].size() : 0; 
    const vector< char > chars1 = {'M', 'A', 'S'};
    const vector<vector< char > > matches2 =  {{'M', 'M', 'S', 'S'}, {'S', 'M', 'M', 'S'}, {'S', 'S', 'M', 'M'}, {'M', 'S', 'S', 'M'}};

    int tot1 = 0, tot2 = 0;
    int di, dj, ii, jj;
    bool flag;
    vector < char > tmp = {};
    for (int i = 0; i < h; i++) {
        for (int j = 0; j < w; j++) {
            if (rows[i][j] == 'X') {
                for (int k = 0; k < directions.size(); k++) {
                    di = directions[k][0],
                    dj = directions[k][1];
                    flag = true;
                    ii = i + di;
                    jj = j + dj;
                    for (int l = 0; l < chars1.size(); l++) {
                        if ((0 <= ii) && (ii < h) && (0 <= jj) && (jj <= w) && rows[ii][jj] == chars1[l]) {
                            ii += di;
                            jj += dj;
                            continue;
                        } else {
                            flag = false;
                            break;
                        }
                    }
                    if (flag) tot1++;
                }
            }
            if (rows[i][j] == 'A') {
                tmp = {};
                for (int k = 0; k < directions2.size(); k++) {
                    di = directions2[k][1];
                    dj = directions2[k][0];
                    ii = i + di;
                    jj = j + dj;
                    if ((0 <= ii) && (ii < h) && (0 <= jj) && (jj < w)) {
                        tmp.push_back(rows[ii][jj]);
                    }
                }
                for (int k = 0; k < matches2.size(); k++) {
                    if (matches2[k] == tmp) {
                        tot2++;
                    }
                }
            }
        }
    }

    cout << "Part 1: " << tot1 << endl;
    cout << "Part 2: " << tot2 << endl;
}